### PR TITLE
Add Automated Provenance Capture to CDMS

### PR DESCRIPTION
**Pull request by chaosphere2112**
This pull request adds in automated provenance capture for variables to CDMS. It generates an directed acyclic graph representing a series of operations on files/variables, and can extract the minimum subtree of that graph to represent the series of operations to create a specific variable. This subtree is capable of being serialized in JSON, which is supported directly in the code. This JSON structure can later be passed back into CDMS, at which point it will attempt to re-enact the derivation of the variable.

This pull request is not yet ready for master, but with a few tweaks it should be ready to start testing for use for VCDAT/VCS.js. I've stubbed in the "convenient" API for using it:

``` python
import cdms2, vcs
# if you don't have the sample data, do vcs.download_sample_data()
f = cdms2.open(vcs.sample_data + "/clt.nc")
s = f['clt']
north = s(latitude=(0, 90))
south = s(latitude=(-90, 0))
import cdms2.MV2
# average across time
n_avgs = cdms2.MV2.average(north, axis=0)
s_avgs = cdms2.MV2.average(south, axis=0)
# calculate diffs
n_s_diffs = n_avgs - s_avgs

# There's an optional "path" argument that will allow export to a file.
json_text = n_s_diffs.export_provenance()
# This should work, though this API is a little wonky (since normally cdms2.open only returns files, and this will return a variable). May change before merging.
derived_n_s_diffs = cdms2.open(json_text)
```

@danlipsa @jbeezley, this is the pull request I mentioned on Wednesday. I ran into a few snags getting it prepped (I had started on the UV-CDAT main repo, and then @doutriaux1 extracted CDMS out to this repo, so I had to jump through some git hoops to follow along).

### CODE REVIEW

**Comment by doutriaux1** [comment]
@chaosphere2112 selectors seem to lose provenace:

``` python
In [2]: s2=s(cdutil.region.AntarcticZone)
In [3]: s2.provenance_node  # None
```

**Comment by doutriaux1** [comment]
@chaosphere2112 we should add tests

File: Lib/provenance/numpy_backend.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+        "ne": cdms2.MV2.not_equal,
+        "not": cdms2.MV2.logical_not,
+        "and": cdms2.MV2.bitwise_and,
+        "or": cdms2.MV2.bitwise_or,
+        "xor": cdms2.MV2.bitwise_xor,
+        "pow": cdms2.MV2.power,
+        "neg": cdms2.MV2.negative,
+    }
+
+    def execute_transform(self, func_name, args, kwargs):
```
Comment (doutriaux1) [inline Lib/provenance/numpy_backend.py:142]: everything in cdms is camelCase

File: Lib/provenance/node.py [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+import logging
+
+
+class Node(object):
+    def __init__(self, backend):
+        self.__cache__ = None
+        self.backend = backend
+
+
+    def get_value(self):
```
Comment (doutriaux1) [inline Lib/provenance/node.py:10]: camelCase?

**Comment by doutriaux1** [comment]
@chaosphere2112 we should make sure everything stays camelCase in cdms.

**Comment by doutriaux1** [comment]
``` python
In [5]: s2=s+3
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-5-b47763f5a24a> in <module>()
----> 1 s2=s+3

/Users/doutriaux1/anaconda2/envs/TEST/lib/python2.7/site-packages/cdms2/avariable.pyc in __add__(self, other)
   1486 
   1487     def __add__(self, other):
-> 1488         return MV.add(self, other)
   1489 
   1490     __radd__ = __add__

/Users/doutriaux1/anaconda2/envs/TEST/lib/python2.7/site-packages/cdms2/MV2.pyc in __call__(self, a, b)
    243 
    244         # Build the provenance node here, retrieve it in other funcs
--> 245         prov_node = track(self.mafunc.__name__, a, b)
    246 
    247         ta = _makeMaskedArg(a)

/Users/doutriaux1/anaconda2/envs/TEST/lib/python2.7/site-packages/cdms2/MV2.pyc in track(name, *args, **kwargs)
     26     if hasattr(parent, "provenance_node"):
     27         operation = parent.track_operation("transform", function=name, **kwargs)
---> 28         parents = [p.provenance_node for p in args]
     29         return parent.track_child(operation, parents)
     30     return None

AttributeError: 'int' object has no attribute 'provenance_node'

In [6]: 
```

**Comment by doutriaux1** [comment]
``` python
In [5]: s2=s+3
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-5-b47763f5a24a> in <module>()
----> 1 s2=s+3

/Users/doutriaux1/anaconda2/envs/TEST/lib/python2.7/site-packages/cdms2/avariable.pyc in __add__(self, other)
   1486 
   1487     def __add__(self, other):
-> 1488         return MV.add(self, other)
   1489 
   1490     __radd__ = __add__

/Users/doutriaux1/anaconda2/envs/TEST/lib/python2.7/site-packages/cdms2/MV2.pyc in __call__(self, a, b)
    243 
    244         # Build the provenance node here, retrieve it in other funcs
--> 245         prov_node = track(self.mafunc.__name__, a, b)
    246 
    247         ta = _makeMaskedArg(a)

/Users/doutriaux1/anaconda2/envs/TEST/lib/python2.7/site-packages/cdms2/MV2.pyc in track(name, *args, **kwargs)
     26     if hasattr(parent, "provenance_node"):
     27         operation = parent.track_operation("transform", function=name, **kwargs)
---> 28         parents = [p.provenance_node for p in args]
     29         return parent.track_child(operation, parents)
     30     return None

AttributeError: 'int' object has no attribute 'provenance_node'

In [6]: s2 = s+3.
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-6-864ab3ae7484> in <module>()
----> 1 s2 = s+3.

/Users/doutriaux1/anaconda2/envs/TEST/lib/python2.7/site-packages/cdms2/avariable.pyc in __add__(self, other)
   1486 
   1487     def __add__(self, other):
-> 1488         return MV.add(self, other)
   1489 
   1490     __radd__ = __add__

/Users/doutriaux1/anaconda2/envs/TEST/lib/python2.7/site-packages/cdms2/MV2.pyc in __call__(self, a, b)
    243 
    244         # Build the provenance node here, retrieve it in other funcs
--> 245         prov_node = track(self.mafunc.__name__, a, b)
    246 
    247         ta = _makeMaskedArg(a)

/Users/doutriaux1/anaconda2/envs/TEST/lib/python2.7/site-packages/cdms2/MV2.pyc in track(name, *args, **kwargs)
     26     if hasattr(parent, "provenance_node"):
     27         operation = parent.track_operation("transform", function=name, **kwargs)
---> 28         parents = [p.provenance_node for p in args]
     29         return parent.track_child(operation, parents)
     30     return None

AttributeError: 'float' object has no attribute 'provenance_node'
```

**Comment by chaosphere2112** [comment]
@doutriaux1 We now support raw values (in practice, anything with that's a valid type in JSON), so this should work better.

**Comment by chaosphere2112** [comment]
This pull request will be superseded by one from a rebased branch once I've completed the provenance functionality.

**Pull request closed without merging**